File: template-repo/templates/nextjs-mysql/lib/ai/AiClient.ts
import 'server-only'

import { aiConfig } from '@/lib/ai/AiConfig'

type AiMessage = {
  role: 'system' | 'user' | 'assistant'
  content: string
}

type ChatOverrides = {
  model?: string
  temperature?: number
  maxTokens?: number
  timeoutSeconds?: number
}

export class AiClient {
  async chat(systemPrompt: string, userMessage: string, overrides: ChatOverrides = {}) {
    // Most business code only needs a single system-plus-user exchange. More
    // advanced workflows can drop down to chatMessages(...) directly.
    return this.chatMessages(
      [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: userMessage }
      ],
      overrides
    )
  }

  async chatMessages(messages: AiMessage[], overrides: ChatOverrides = {}) {
    // Expose a string-in, string-out helper so generated server code can treat
    // AI calls like any other service dependency.
    const response = await this.request('/chat/completions', this.buildBody(messages, false, overrides))
    const payload = (await response.json()) as {
      choices?: Array<{ message?: { content?: string } }>
    }

    return payload.choices?.[0]?.message?.content ?? ''
  }

  async *chatStream(
    systemPrompt: string,
    userMessage: string,
    overrides: ChatOverrides = {}
  ): AsyncGenerator<string, void, void> {
    yield* this.chatMessagesStream(
      [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: userMessage }
      ],
      overrides
    )
  }

  async *chatMessagesStream(
    messages: AiMessage[],
    overrides: ChatOverrides = {}
  ): AsyncGenerator<string, void, void> {
    const response = await this.request('/chat/completions', this.buildBody(messages, true, overrides))
    const reader = response.body?.getReader()
    if (!reader) {
      throw new Error('AI stream response body is not readable')
    }

    const decoder = new TextDecoder()
    let buffer = ''

    while (true) {
      const { done, value } = await reader.read()
      if (done) {
        break
      }

      buffer += decoder.decode(value, { stream: true })
      const lines = buffer.split(/\r?\n/)
      buffer = lines.pop() ?? ''

      for (const rawLine of lines) {
        const line = rawLine.trim()
        // Streaming APIs send control frames alongside text frames. Downstream
        // code only needs the assistant delta content.
        if (!line.startsWith('data: ') || line === 'data: [DONE]') {
          continue
        }

        try {
          const payload = JSON.parse(line.slice(6)) as {
            choices?: Array<{ delta?: { content?: string } }>
          }
          const content = payload.choices?.[0]?.delta?.content ?? ''
          if (content) {
            yield content
          }
        } catch {
          continue
        }
      }
    }
  }

  private buildBody(messages: AiMessage[], stream: boolean, overrides: ChatOverrides) {
    // Centralize request shaping so future features like JSON mode or tool
    // calling can be added without changing every caller.
    const body: Record<string, unknown> = {
      model: overrides.model ?? aiConfig.model,
      messages,
      stream,
      temperature: overrides.temperature ?? aiConfig.temperature
    }

    const maxTokens = overrides.maxTokens ?? aiConfig.maxTokens
    if (maxTokens > 0) {
      body.max_tokens = maxTokens
    }

    return body
  }

  private async request(path: string, body: Record<string, unknown>) {
    const controller = new AbortController()
    const timeout = setTimeout(() => controller.abort(), (aiConfig.timeoutSeconds ?? 60) * 1000)

    try {
      // Route all outbound calls through one helper so timeout handling, auth,
      // and provider compatibility stay consistent across generated features.
      const response = await fetch(`${aiConfig.baseUrl.replace(/\/+$/, '')}${path}`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${aiConfig.apiKey}`
        },
        body: JSON.stringify(body),
        signal: controller.signal,
        cache: 'no-store'
      })

      if (!response.ok) {
        throw new Error(`AI API error ${response.status}: ${await response.text()}`)
      }

      return response
    } finally {
      clearTimeout(timeout)
    }
  }
}

export const aiClient = new AiClient()

File: template-repo/templates/springboot-vue3-mysql/backend/src/main/java/com/smartark/template/ai/AiClient.java
package com.smartark.template.ai;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import org.springframework.stereotype.Component;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.List;
import java.util.function.Consumer;

/**
 * Unified LLM client supporting OpenAI-compatible APIs.
 *
 * Usage in any service:
 *   String result = aiClient.chat("Analyze this resume", resumeText);
 *   aiClient.chatStream("Optimize this", resumeText, chunk -> emitter.send(chunk));
 *
 * The client exposes both a simple single-turn helper and a lower-level
 * message-list overload. Generated business code can start simple and move to
 * multi-step orchestration without introducing a second AI abstraction.
 */
@Component
public class AiClient {

    private final AiConfig config;
    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;

    public AiClient(AiConfig config, ObjectMapper objectMapper) {
        this.config = config;
        this.objectMapper = objectMapper;
        this.httpClient = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(10))
                .build();
    }

    /** Single-turn chat helper. */
    public String chat(String systemPrompt, String userMessage) {
        return chat(List.of(
                new Message("system", systemPrompt),
                new Message("user", userMessage)
        ));
    }

    /** Multi-turn chat with full message list. */
    public String chat(List<Message> messages) {
        try {
            // Build the payload in one place so sync and streaming paths use the
            // same request shape and stay easy to reason about.
            String body = buildBody(messages, false);
            HttpResponse<String> response = httpClient.send(buildRequest(body), HttpResponse.BodyHandlers.ofString());
            if (response.statusCode() != 200) {
                throw new RuntimeException("AI API error " + response.statusCode() + ": " + response.body());
            }
            JsonNode root = objectMapper.readTree(response.body());
            return root.at("/choices/0/message/content").asText("");
        } catch (RuntimeException e) {
            throw e;
        } catch (Exception e) {
            throw new RuntimeException("AI chat failed: " + e.getMessage(), e);
        }
    }

    /** Streaming chat helper. */
    public void chatStream(String systemPrompt, String userMessage, Consumer<String> onChunk) {
        chatStream(List.of(
                new Message("system", systemPrompt),
                new Message("user", userMessage)
        ), onChunk);
    }

    /** Streaming chat with full message list. */
    public void chatStream(List<Message> messages, Consumer<String> onChunk) {
        try {
            String body = buildBody(messages, true);
            HttpResponse<java.io.InputStream> response = httpClient.send(
                    buildRequest(body),
                    HttpResponse.BodyHandlers.ofInputStream()
            );
            if (response.statusCode() != 200) {
                throw new RuntimeException("AI API error " + response.statusCode());
            }
            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(response.body(), StandardCharsets.UTF_8))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    // Streaming providers emit one SSE line at a time. We only
                    // forward meaningful text deltas to the caller.
                    if (line.startsWith("data: ") && !line.equals("data: [DONE]")) {
                        JsonNode chunk = objectMapper.readTree(line.substring(6));
                        String content = chunk.at("/choices/0/delta/content").asText("");
                        if (!content.isEmpty()) {
                            onChunk.accept(content);
                        }
                    }
                }
            }
        } catch (RuntimeException e) {
            throw e;
        } catch (Exception e) {
            throw new RuntimeException("AI stream failed: " + e.getMessage(), e);
        }
    }

    private String buildBody(List<Message> messages, boolean stream) {
        try {
            ObjectNode root = objectMapper.createObjectNode();
            root.put("model", config.getModel());
            root.put("stream", stream);
            root.put("temperature", config.getTemperature());
            if (config.getMaxTokens() > 0) {
                root.put("max_tokens", config.getMaxTokens());
            }
            // Preserve message order exactly so callers can model multi-step
            // prompt flows without the client reinterpreting conversation state.
            ArrayNode messageArray = root.putArray("messages");
            for (Message message : messages) {
                messageArray.addObject()
                        .put("role", message.role())
                        .put("content", message.content());
            }
            return objectMapper.writeValueAsString(root);
        } catch (Exception e) {
            throw new RuntimeException("Failed to build AI request body", e);
        }
    }

    private HttpRequest buildRequest(String body) {
        // The template keeps provider integration OpenAI-compatible so switching
        // from one gateway to another only requires config changes.
        String url = config.getBaseUrl().replaceAll("/+$", "") + "/chat/completions";
        return HttpRequest.newBuilder()
                .uri(URI.create(url))
                .header("Content-Type", "application/json")
                .header("Authorization", "Bearer " + config.getApiKey())
                .timeout(Duration.ofSeconds(config.getTimeoutSeconds()))
                .POST(HttpRequest.BodyPublishers.ofString(body))
                .build();
    }

    public record Message(String role, String content) {}
}
